File: packages/shared/src/condition.ts
// 조건(ABAC/CEL) 트리 계약 (lazyfga-13).
// WAF식 And/Or 블록 빌더가 만들고, lazyfga-14가 OpenFGA CEL로 컴파일한다.
// 지원 피연산자는 전부 OpenFGA 네이티브 condition 파라미터 타입에 대응한다 — lazyFGA는
// 조건을 직접 평가하지 않고, CEL 생성·선언만 하며 평가는 OpenFGA가 Check 시 수행한다.
import { z } from "zod";
import { CEL_RESERVED, IDENT_RE, RESERVED_WORDS } from "./ident";

/** 조건 파라미터 타입. OpenFGA 네이티브 condition 파라미터 타입의 부분집합(MVP). */
export type ConditionParamType = "timestamp" | "ipaddress" | "string" | "int" | "double" | "bool";

export interface ConditionParam {
  /** CEL 파라미터 이름. 예: "current_time", "user_ip" */
  name: string;
  type: ConditionParamType;
}

/** 시간 비교의 우변: 고정 시각(literal) 또는 다른 timestamp 파라미터. */
export type TimeRhs =
  | { kind: "literal"; rfc3339: string }
  | { kind: "param"; param: string };

/** 단일 비교(leaf 술어). */
export type ConditionLeaf =
  | { kind: "time"; param: string; op: "lt" | "lte" | "gt" | "gte"; rhs: TimeRhs }
  | { kind: "ip"; param: string; op: "in_cidr"; cidr: string }
  | {
      kind: "value";
      param: string;
      op: "eq" | "neq" | "lt" | "lte" | "gt" | "gte";
      value: string | number | boolean;
    };

/** AND/OR 그룹(WAF식). children는 leaf 또는 중첩 그룹. */
export interface ConditionGroup {
  op: "and" | "or";
  children: ConditionNode[];
}
export type ConditionNode = ConditionGroup | ConditionLeaf;

/** 이름 붙은 재사용 조건. OpenFGA `condition <name>(<params>) { <CEL> }` 한 블록에 대응. */
export interface ConditionDef {
  name: string;
  params: ConditionParam[];
  tree: ConditionNode;
}

// ── 런타임 스키마(zod) ────────────────────────────────────────────────────────

const conditionParamTypeSchema = z.enum([
  "timestamp",
  "ipaddress",
  "string",
  "int",
  "double",
  "bool",
]);

const conditionParamSchema: z.ZodType<ConditionParam> = z.object({
  name: z.string(),
  type: conditionParamTypeSchema,
});

const timeRhsSchema: z.ZodType<TimeRhs> = z.discriminatedUnion("kind", [
  z.object({ kind: z.literal("literal"), rfc3339: z.string() }),
  z.object({ kind: z.literal("param"), param: z.string() }),
]);

const conditionLeafSchema: z.ZodType<ConditionLeaf> = z.discriminatedUnion("kind", [
  z.object({
    kind: z.literal("time"),
    param: z.string(),
    op: z.enum(["lt", "lte", "gt", "gte"]),
    rhs: timeRhsSchema,
  }),
  z.object({
    kind: z.literal("ip"),
    param: z.string(),
    op: z.literal("in_cidr"),
    cidr: z.string(),
  }),
  z.object({
    kind: z.literal("value"),
    param: z.string(),
    op: z.enum(["eq", "neq", "lt", "lte", "gt", "gte"]),
    value: z.union([z.string(), z.number(), z.boolean()]),
  }),
]);

const conditionNodeSchema: z.ZodType<ConditionNode> = z.lazy(() =>
  z.union([
    z.object({ op: z.enum(["and", "or"]), children: z.array(conditionNodeSchema) }),
    conditionLeafSchema,
  ]),
);

export const conditionDefSchema: z.ZodType<ConditionDef> = z.object({
  name: z.string(),
  params: z.array(conditionParamSchema),
  tree: conditionNodeSchema,
});

// ── 사람이 읽는 미리보기(describeCondition) ──────────────────────────────────────

const isGroup = (n: ConditionNode): n is ConditionGroup => "children" in n;

const TIME_OP: Record<"lt" | "lte" | "gt" | "gte", string> = {
  lt: "<",
  lte: "<=",
  gt: ">",
  gte: ">=",
};
const VALUE_OP: Record<"eq" | "neq" | "lt" | "lte" | "gt" | "gte", string> = {
  eq: "==",
  neq: "!=",
  lt: "<",
  lte: "<=",
  gt: ">",
  gte: ">=",
};

function describeLeaf(leaf: ConditionLeaf): string {
  if (leaf.kind === "time") {
    const rhs = leaf.rhs.kind === "literal" ? leaf.rhs.rfc3339 : leaf.rhs.param;
    return `${leaf.param} ${TIME_OP[leaf.op]} ${rhs}`;
  }
  if (leaf.kind === "ip") return `${leaf.param} in ${leaf.cidr}`;
  const v = typeof leaf.value === "string" ? `"${leaf.value}"` : String(leaf.value);
  return `${leaf.param} ${VALUE_OP[leaf.op]} ${v}`;
}

/**
 * 조건 트리를 사람이 읽는 한 줄로 렌더(순수, 결정적).
 * child가 2개 이상인 그룹은 괄호로 감싸 and/or 우선순위 모호성을 제거한다.
 */
export function describeCondition(node: ConditionNode): string {
  if (!isGroup(node)) return describeLeaf(node);
  if (node.children.length === 0) return "(empty)";
  if (node.children.length === 1) return describeCondition(node.children[0]!);
  const sep = node.op === "and" ? " AND " : " OR ";
  return `(${node.children.map(describeCondition).join(sep)})`;
}

// ── 정적 검증(validateConditionDef) ───────────────────────────────────────────

export type ConditionErrorCode =
  | "BAD_NAME"
  | "DUP_PARAM"
  | "UNKNOWN_PARAM"
  | "TYPE_MISMATCH"
  | "BAD_CIDR"
  | "BAD_TIMESTAMP"
  | "EMPTY_GROUP";

export interface ConditionError {
  code: ConditionErrorCode;
  path: string;
  message: string;
}

const RFC3339_RE = /^\d{4}-\d{2}-\d{2}[Tt]\d{2}:\d{2}:\d{2}(\.\d+)?([Zz]|[+-]\d{2}:\d{2})$/;

/** IPv4/IPv6 CIDR(addr/prefix) 형식 검사(보수적). */
function isCidr(s: string): boolean {
  const slash = s.lastIndexOf("/");
  if (slash < 0) return false;
  const addr = s.slice(0, slash);
  const prefixStr = s.slice(slash + 1);
  if (!/^\d+$/.test(prefixStr)) return false;
  const prefix = Number(prefixStr);
  if (addr.includes(":")) {
    // IPv6(보수적): hex 그룹 + 콜론(IPv4-mapped 위해 점 허용), prefix 0~128.
    return prefix >= 0 && prefix <= 128 && /^[0-9a-fA-F:.]+$/.test(addr) && addr.length > 0;
  }
  const octets = addr.split(".");
  if (octets.length !== 4) return false;
  if (!octets.every((o) => /^\d{1,3}$/.test(o) && Number(o) <= 255)) return false;
  return prefix >= 0 && prefix <= 32;
}

const VALUE_TYPES: ReadonlySet<ConditionParamType> = new Set([
  "string",
  "int",
  "double",
  "bool",
]);
const ORDER_OPS: ReadonlySet<string> = new Set(["lt", "lte", "gt", "gte"]);

/** 조건 정의의 정적 검증(빈 배열 = 유효, 예외 없음). lazyfga-13 §4.3 규칙 1~7. */
export function validateConditionDef(def: ConditionDef): ConditionError[] {
  const errors: ConditionError[] = [];
  const add = (code: ConditionErrorCode, path: string, message: string): void => {
    errors.push({ code, path, message });
  };
  // condition/param 이름은 CEL 식으로 흘러가므로 CEL 예약어도 금지(상수식/문법파손 방지, lazyfga-14 hardening).
  const badIdent = (name: string): boolean =>
    !IDENT_RE.test(name) || RESERVED_WORDS.has(name) || CEL_RESERVED.has(name);

  // rule 1: 조건 이름.
  if (badIdent(def.name)) add("BAD_NAME", "name", `invalid condition name: "${def.name}"`);

  // params: rule 1(이름) + rule 2(유일).
  const paramType = new Map<string, ConditionParamType>();
  const seen = new Set<string>();
  def.params.forEach((p, i) => {
    if (badIdent(p.name)) add("BAD_NAME", `params[${i}].name`, `invalid param name: "${p.name}"`);
    if (seen.has(p.name)) add("DUP_PARAM", `params[${i}].name`, `duplicate param: "${p.name}"`);
    seen.add(p.name);
    paramType.set(p.name, p.type);
  });

  const expectParam = (param: string, path: string, wanted: ConditionParamType[]): void => {
    if (!paramType.has(param)) {
      add("UNKNOWN_PARAM", `${path}.param`, `unknown param: "${param}"`);
      return;
    }
    const t = paramType.get(param)!;
    if (!wanted.includes(t)) {
      add("TYPE_MISMATCH", `${path}.param`, `param "${param}" is ${t}, expected ${wanted.join("|")}`);
    }
  };

  const visitLeaf = (leaf: ConditionLeaf, path: string): void => {
    if (leaf.kind === "time") {
      expectParam(leaf.param, path, ["timestamp"]);
      if (leaf.rhs.kind === "literal") {
        if (!RFC3339_RE.test(leaf.rhs.rfc3339)) {
          add("BAD_TIMESTAMP", `${path}.rhs.rfc3339`, `invalid RFC3339: "${leaf.rhs.rfc3339}"`);
        }
      } else if (!paramType.has(leaf.rhs.param)) {
        add("UNKNOWN_PARAM", `${path}.rhs.param`, `unknown param: "${leaf.rhs.param}"`);
      } else if (paramType.get(leaf.rhs.param) !== "timestamp") {
        add("TYPE_MISMATCH", `${path}.rhs.param`, `param "${leaf.rhs.param}" must be timestamp`);
      }
    } else if (leaf.kind === "ip") {
      expectParam(leaf.param, path, ["ipaddress"]);
      if (!isCidr(leaf.cidr)) add("BAD_CIDR", `${path}.cidr`, `invalid CIDR: "${leaf.cidr}"`);
    } else {
      expectParam(leaf.param, path, ["string", "int", "double", "bool"]);
      const t = paramType.get(leaf.param);
      if (t !== undefined && VALUE_TYPES.has(t)) {
        const v = leaf.value;
        const ok =
          (t === "string" && typeof v === "string") ||
          (t === "int" && typeof v === "number" && Number.isSafeInteger(v)) ||
          (t === "double" && typeof v === "number" && Number.isFinite(v)) ||
          (t === "bool" && typeof v === "boolean");
        if (!ok) {
          add("TYPE_MISMATCH", `${path}.value`, `value ${JSON.stringify(v)} does not match param type ${t}`);
        }
        if (t === "bool" && ORDER_OPS.has(leaf.op)) {
          add("TYPE_MISMATCH", `${path}.op`, `ordering op "${leaf.op}" not allowed on bool param`);
        }
      }
    }
  };

  const visit = (node: ConditionNode, path: string): void => {
    if (isGroup(node)) {
      if (node.children.length === 0) add("EMPTY_GROUP", path, "group must have >= 1 child");
      node.children.forEach((c, i) => visit(c, `${path}.children[${i}]`));
      return;
    }
    visitLeaf(node, path);
  };

  visit(def.tree, "tree");
  return errors;
}

File: apps/api/internal/contract/condition.go
package contract

import (
	"bytes"
	"encoding/json"
	"errors"
	"math"

	"github.com/antegral-constellation/lazyfga/api/internal/jsutil"
)

// 조건(ABAC/CEL) 트리 계약(LFGA-13/14). 판별 유니온(TimeRhs/ConditionLeaf/ConditionNode)과
// 값 리터럴(string|number|bool)은 커스텀 (un)marshaler로 TS와 바이트 호환 JSON을 유지한다.

// ConditionParamType: OpenFGA 네이티브 condition 파라미터 타입의 부분집합(MVP).
// "timestamp" | "ipaddress" | "string" | "int" | "double" | "bool".

// ConditionParam은 CEL 파라미터(name/type).
type ConditionParam struct {
	Name string `json:"name"`
	Type string `json:"type"`
}

// ValueKind는 ConditionValue 판별자.
type ValueKind int

const (
	ValueString ValueKind = iota
	ValueNumber
	ValueBool
)

// ConditionValue는 value leaf의 RHS(string|number|bool). 어떤 JSON 타입이었는지 보존한다.
type ConditionValue struct {
	Kind ValueKind
	Str  string
	Num  float64
	Bool bool
}

// StringValue/NumberValue/BoolValue: 생성 헬퍼.
func StringValue(s string) ConditionValue  { return ConditionValue{Kind: ValueString, Str: s} }
func NumberValue(f float64) ConditionValue { return ConditionValue{Kind: ValueNumber, Num: f} }
func BoolValue(b bool) ConditionValue      { return ConditionValue{Kind: ValueBool, Bool: b} }

func (v ConditionValue) MarshalJSON() ([]byte, error) {
	switch v.Kind {
	case ValueString:
		// JS JSON.stringify와 바이트 동일(\b/\f 축약, <>&·U+2028/29 raw 포함) — jsutil이 단일 원본.
		return []byte(jsutil.JSONString(v.Str)), nil
	case ValueNumber:
		// JS 숫자 표기(-0→"0", 지수 무패딩, [1e-6,1e21) 십진). 비유한값은 JSON이 아니므로 오류.
		if math.IsNaN(v.Num) || math.IsInf(v.Num, 0) {
			return nil, errors.New("contract: condition value number must be finite")
		}
		return []byte(jsutil.NumberString(v.Num)), nil
	case ValueBool:
		if v.Bool {
			return []byte("true"), nil
		}
		return []byte("false"), nil
	default:
		return nil, errors.New("contract: invalid ConditionValue kind")
	}
}

func (v *ConditionValue) UnmarshalJSON(b []byte) error {
	trimmed := bytes.TrimSpace(b)
	if len(trimmed) == 0 || bytes.Equal(trimmed, []byte("null")) {
		return errors.New("contract: condition value must be string|number|bool")
	}
	switch trimmed[0] {
	case '"':
		var s string
		if err := json.Unmarshal(trimmed, &s); err != nil {
			return err
		}
		*v = StringValue(s)
	case 't', 'f':
		var x bool
		if err := json.Unmarshal(trimmed, &x); err != nil {
			return err
		}
		*v = BoolValue(x)
	default:
		var f float64
		if err := json.Unmarshal(trimmed, &f); err != nil {
			return err
		}
		*v = NumberValue(f)
	}
	return nil
}

// TimeRhs: 시간 비교 우변 — literal(rfc3339) 또는 param.
type TimeRhs struct {
	Kind    string // "literal" | "param"
	RFC3339 string // literal
	Param   string // param
}

func (r TimeRhs) MarshalJSON() ([]byte, error) {
	switch r.Kind {
	case "literal":
		return marshalNoEscape(struct {
			Kind    string `json:"kind"`
			RFC3339 string `json:"rfc3339"`
		}{r.Kind, r.RFC3339})
	case "param":
		return marshalNoEscape(struct {
			Kind  string `json:"kind"`
			Param string `json:"param"`
		}{r.Kind, r.Param})
	default:
		return nil, errors.New("contract: invalid TimeRhs kind")
	}
}

func (r *TimeRhs) UnmarshalJSON(b []byte) error {
	var probe struct {
		Kind    string `json:"kind"`
		RFC3339 string `json:"rfc3339"`
		Param   string `json:"param"`
	}
	if err := json.Unmarshal(b, &probe); err != nil {
		return err
	}
	switch probe.Kind {
	case "literal":
		*r = TimeRhs{Kind: "literal", RFC3339: probe.RFC3339}
	case "param":
		*r = TimeRhs{Kind: "param", Param: probe.Param}
	default:
		return errors.New("contract: unknown TimeRhs kind: " + probe.Kind)
	}
	return nil
}

// ConditionLeaf: 단일 비교 leaf(time|ip|value).
type ConditionLeaf struct {
	Kind  string // "time" | "ip" | "value"
	Param string
	Op    string
	Rhs   *TimeRhs        // time
	Cidr  string          // ip
	Value *ConditionValue // value
}

func (l ConditionLeaf) MarshalJSON() ([]byte, error) {
	switch l.Kind {
	case "time":
		return marshalNoEscape(struct {
			Kind  string  `json:"kind"`
			Param string  `json:"param"`
			Op    string  `json:"op"`
			Rhs   TimeRhs `json:"rhs"`
		}{l.Kind, l.Param, l.Op, deref(l.Rhs)})
	case "ip":
		return marshalNoEscape(struct {
			Kind  string `json:"kind"`
			Param string `json:"param"`
			Op    string `json:"op"`
			Cidr  string `json:"cidr"`
		}{l.Kind, l.Param, l.Op, l.Cidr})
	case "value":
		return marshalNoEscape(struct {
			Kind  string         `json:"kind"`
			Param string         `json:"param"`
			Op    string         `json:"op"`
			Value ConditionValue `json:"value"`
		}{l.Kind, l.Param, l.Op, derefValue(l.Value)})
	default:
		return nil, errors.New("contract: invalid ConditionLeaf kind")
	}
}

func (l *ConditionLeaf) UnmarshalJSON(b []byte) error {
	var probe struct {
		Kind  string          `json:"kind"`
		Param string          `json:"param"`
		Op    string          `json:"op"`
		Rhs   *TimeRhs        `json:"rhs"`
		Cidr  string          `json:"cidr"`
		Value *ConditionValue `json:"value"`
	}
	if err := json.Unmarshal(b, &probe); err != nil {
		return err
	}
	switch probe.Kind {
	case "time":
		*l = ConditionLeaf{Kind: "time", Param: probe.Param, Op: probe.Op, Rhs: probe.Rhs}
	case "ip":
		*l = ConditionLeaf{Kind: "ip", Param: probe.Param, Op: probe.Op, Cidr: probe.Cidr}
	case "value":
		*l = ConditionLeaf{Kind: "value", Param: probe.Param, Op: probe.Op, Value: probe.Value}
	default:
		return errors.New("contract: unknown ConditionLeaf kind: " + probe.Kind)
	}
	return nil
}

// ConditionGroup: AND/OR 그룹.
type ConditionGroup struct {
	Op       string          `json:"op"`
	Children []ConditionNode `json:"children"`
}

// ConditionNode: group 또는 leaf(둘 중 하나만 non-nil). 판별은 "children" 키 유무.
type ConditionNode struct {
	Group *ConditionGroup
	Leaf  *ConditionLeaf
}

// GroupNode/LeafNode: 생성 헬퍼.
func GroupNode(op string, children ...ConditionNode) ConditionNode {
	return ConditionNode{Group: &ConditionGroup{Op: op, Children: children}}
}
func LeafNode(l ConditionLeaf) ConditionNode { return ConditionNode{Leaf: &l} }

func (n ConditionNode) isGroup() bool { return n.Group != nil }

func (n ConditionNode) MarshalJSON() ([]byte, error) {
	if n.Group != nil {
		return marshalNoEscape(n.Group)
	}
	if n.Leaf != nil {
		return marshalNoEscape(*n.Leaf)
	}
	return nil, errors.New("contract: empty ConditionNode")
}

func (n *ConditionNode) UnmarshalJSON(b []byte) error {
	var probe map[string]json.RawMessage
	if err := json.Unmarshal(b, &probe); err != nil {
		return err
	}
	// zod 순서 union 재현: group(op∈and|or + children 배열)이 성공하면 group, 실패하면 leaf 폴백.
	// children 키 존재만으로 group을 강제하지 않는다(잉여 children을 가진 유효 leaf 수용).
	if opRaw, hasOp := probe["op"]; hasOp {
		var op string
		if json.Unmarshal(opRaw, &op) == nil && (op == "and" || op == "or") {
			if _, hasChildren := probe["children"]; hasChildren {
				var g ConditionGroup
				if err := json.Unmarshal(b, &g); err == nil {
					*n = ConditionNode{Group: &g}
					return nil
				}
			}
		}
	}
	var l ConditionLeaf
	if err := json.Unmarshal(b, &l); err != nil {
		return err
	}
	*n = ConditionNode{Leaf: &l}
	return nil
}

// ConditionDef: 이름 붙은 재사용 조건(OpenFGA condition 블록 1개에 대응).
type ConditionDef struct {
	Name   string           `json:"name"`
	Params []ConditionParam `json:"params"`
	Tree   ConditionNode    `json:"tree"`
}

func deref(r *TimeRhs) TimeRhs {
	if r == nil {
		return TimeRhs{}
	}
	return *r
}

func derefValue(v *ConditionValue) ConditionValue {
	if v == nil {
		return ConditionValue{}
	}
	return *v
}
